psa-crypto-sys/build.rs

```diff
@@ -66,6 +66,73 @@ mod common {
     use std::io::{Error, ErrorKind, Result};
     use std::path::{Path, PathBuf};
 
+    #[cfg(any(feature = "prefix", feature = "operations"))]
+    pub fn get_external_mbedtls() -> Option<Result<(String, String)>> {
+        if env::var("MBEDTLS_LIB_DIR").is_err() ^ env::var("MBEDTLS_INCLUDE_DIR").is_err() {
+            return Some(Err(Error::new(
+                ErrorKind::Other,
+                "both environment variables MBEDTLS_LIB_DIR and MBEDTLS_INCLUDE_DIR need to be set for operations feature",
+            )));
+        }
+
+        if let (Ok(lib_dir), Ok(include_dir)) =
+            (env::var("MBEDTLS_LIB_DIR"), env::var("MBEDTLS_INCLUDE_DIR"))
+        {
+            println!("Found environment variables, using external MbedTLS");
+            return Some(Ok((include_dir, lib_dir)));
+        }
+
+        if let Ok(mbedtls_result) = pkg_config::Config::new()
+            .range_version("3.5".."4.0")
+            .probe("mbedtls")
+        {
+            let include_dirs: Vec<String> = mbedtls_result
+                .include_paths
+                .into_iter()
+                .map(|x: PathBuf| -> String { x.into_os_string().into_string().unwrap() })
+                .collect();
+            let include_dir = include_dirs.join(" ");
+            // The current build framework doesn't support multiple lib paths for -L unfortuantely, so
+            // we just take the first element, which is enough for now :-(
+            let lib_dir = <PathBuf as Clone>::clone(&mbedtls_result.link_paths[0])
+                .into_os_string()
+                .into_string()
+                .unwrap();
+            println!("Found pkg-config mbedtls, using external MbedTLS");
+            return Some(Ok((include_dir, lib_dir)));
+        }
+
+        // No env vars set and no discovered package through pkg-config
+        None
+    }
+
+    #[cfg(all(feature = "interface", not(feature = "operations")))]
+    pub fn get_external_mbedtls_include_only() -> Result<String> {
+        if let Ok(include_dir) = env::var("MBEDTLS_INCLUDE_DIR") {
+            println!("Found environment variable, using external MbedTLS");
+            return Ok(include_dir);
+        }
+
+        if let Ok(mbedtls_result) = pkg_config::Config::new()
+            .range_version("3.5".."4.0")
+            .probe("mbedtls")
+        {
+            let include_dirs: Vec<String> = mbedtls_result
+                .include_paths
+                .into_iter()
+                .map(|x: PathBuf| -> String { x.into_os_string().into_string().unwrap() })
+                .collect();
+            let include_dir = include_dirs.join(" ");
+
+            return Ok(include_dir);
+        }
+
+        Err(Error::new(
+            ErrorKind::Other,
+            "interface feature necessitates MBEDTLS_INCLUDE_DIR environment variable",
```

Comment:
Should we also include `or library discoverable by pkg-config` (or something along similar lines) in the error log here?

Reply:
No, this is for the interface feature which doesn't need the libdir, just the headers.